src/HexagonAlignLoads.cpp

```diff
@@ -0,0 +1,192 @@
+                        // We know nothing about alignment. Give up.
+                        // TODO: Fix this.
+                        debug(4) << "HexagonAlignLoads: Cannot reason about alignment.\n";
+                        debug(4) << "HexagonAlignLoads: Type: " << op->type << "\n";
+                        debug(4) << "HexagonAlignLoads: Index: " << index << "\n";
+                        expr = op;
+                        return;
+                    }
+                    int base_mod = base_lanes_off + mod_imp(mod_rem.modulus, lanes);
+                    int rem_mod = mod_imp(mod_rem.remainder, lanes);
+                    if (!(base_mod + rem_mod)) {
+                        expr = op;
+                        debug(4) << "HexagonAlignLoads: Encountered a perfectly aligned load.\n";
+                        debug(4) << "HexagonAlignLoads: Type: " << op->type << "\n";
+                        debug(4) << "HexagonAlignLoads: Index: " << index << "\n";
+                        return;
+                    } else {
+                        Expr base = ramp->base;
+                        const Add *add = base.as<Add>();
+                        const IntImm *b = add->b.as<IntImm>();
+                        if (!b) {
+                            // Should we be expecting the index to be in simplified
+                            // canonical form, i.e. expression + IntImm?
+                            debug(4) << "HexagonAlignLoads: add->b is not a constant\n";
+                            debug(4) << "HexagonAlignLoads: Type: " << op->type << "\n";
+                            debug(4) << "HexagonAlignLoads: Index: " << index << "\n";
+                            expr = op;
+                            return;
+                        }
+                        Expr base_low = base - (base_mod + rem_mod);
+                        Expr ramp_low = Ramp::make(simplify(base_low), 1, lanes);
+                        Expr ramp_high = Ramp::make(simplify(base_low + lanes), 1, lanes);
+                        Expr load_low = Load::make(op->type, op->name, ramp_low, op->image, op->param);
+                        Expr load_high = Load::make(op->type, op->name, ramp_high, op->image, op->param);
+                        // slice_vector considers the vectors in it to be concatenated and the result
+                        // is a vector containing as many lanes as the last argument and the vector
+                        // begins at lane number specified by the penultimate arg.
+                        // So, slice_vector(2, a, b, 126, 128) gives a vector of 128 lanes starting
+                        // from lanes 126 through 253 of the concatenated vector a.b.
+                        debug(4) << "HexagonAlignLoads: Unaligned Load: Converting " << (Expr) op << " into ...\n";
+                        expr = Call::make(op->type, Call::slice_vector, { make_two(Int(32)), load_low, load_high, make_const(Int(32), (base_mod + rem_mod)),
```

Comment:
For 32 bit int arguments, you can just pass integers as Exprs. It would be less code noise to just do that.